atomic_reactor/plugins/post_koji_upload.py

```diff
@@ -0,0 +1,642 @@
+        'docker save' output to upload.
+
+        For metadata-only builds, an empty file is used instead of the
+        output of 'docker save'.
+
+        :param arch: str, architecture for this output
+        :return: tuple, (metadata dict, Output instance)
+
+        """
+
+        image_id = self.workflow.builder.image_id
+        saved_image = self.workflow.exported_image_sequence[-1].get('path')
+        ext = saved_image.split('.', 1)[1]
+        name_fmt = 'docker-image-{id}.{arch}.{ext}'
+        image_name = name_fmt.format(id=image_id, arch=arch, ext=ext)
+        metadata = self.get_output_metadata(saved_image, image_name)
+        output = Output(file=open(saved_image), metadata=metadata)
+
+        return metadata, output
+
+    def get_digests(self):
+        """
+        Returns a map of repositories to digests
+        """
+
+        digests = {}  # repository -> digest
+        for registry in self.workflow.push_conf.docker_registries:
+            for image in self.workflow.tag_conf.images:
+                image_str = image.to_str()
+                if image_str in registry.digests:
+                    # pulp/crane supports only manifest schema v1
+                    if self.workflow.push_conf.pulp_registries:
+                        digest = registry.digests[image_str].v1
+                    else:
+                        digest = registry.digests[image_str].default
+                    digests[image.to_str(registry=False)] = digest
+
+        return digests
+
+    def get_repositories(self, digests):
+        """
+        Build the repositories metadata
+
+        :param digests: dict, repository -> digest
+        """
+        if self.workflow.push_conf.pulp_registries:
+            # If pulp was used, only report pulp images
+            registries = self.workflow.push_conf.pulp_registries
+        else:
+            # Otherwise report all the images we pushed
+            registries = self.workflow.push_conf.all_registries
+
+        output_images = []
+        for registry in registries:
+            image = self.pullspec_image.copy()
+            image.registry = registry.uri
+            pullspec = image.to_str()
+
+            output_images.append(pullspec)
+
+            digest = digests.get(image.to_str(registry=False))
+            if digest:
+                digest_pullspec = image.to_str(tag=False) + "@" + digest
+                output_images.append(digest_pullspec)
+
+        return output_images
+
+    def get_output(self, buildroot_id):
+        """
+        Build the 'output' section of the metadata.
+
+        :return: list, Output instances
+        """
+
+        def add_buildroot_id(output):
+            logfile, metadata = output
+            metadata.update({'buildroot_id': buildroot_id})
+            return Output(file=logfile, metadata=metadata)
+
+        def add_log_type(output):
+            logfile, metadata = output
+            metadata.update({'type': 'log', 'arch': 'noarch'})
```

Comment:
We should move the `arch = os.uname()[4]` before this line and use it to set `'arch': arch` instead of 'noarch'.
http://osbs.readthedocs.io/en/latest/multiarch.html#koji-build